sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -697,7 +697,12 @@ class CodegenContext {
           }
         """
       s"${addNewFunction(compareFunc, funcCode)}($c1, $c2)"
-    case other if other.isInstanceOf[AtomicType] => s"$c1.compare($c2)"
+    case other if other.isInstanceOf[AtomicType] =>
+      if (other.getClass.getMethods.map(_.getName).filter(_ matches "compare").size == 1) {
```

Comment:
This is going to get a `compare` method from the given data type. It is not the actual runtime object.

Reply:
@viirya thats a good point.will rethink.